aip/general/0129.md

```diff
@@ -0,0 +1,144 @@
+---
+id: 129
+state: draft
+created: 2023-09-20
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-Modified Values and Defaults
+
+Services often provide default values for resource fields, and occasionally normalize the user input before returning it in the response. The guidance herein describes how services document such behavior for the benefit of consumers.
+
+## Guidance
+
+### Single Owner Fields
+
+Fields **must** have a single owner, whether that is the client or the server.
+Server owned fields **must** be indicated with the `OUTPUT_ONLY` field_behavior.
+All other types of fields **must** be considered to be owned by the client. The
+server **must** respect the value (or lack thereof) for all client owned fields
+and not modify them.
+
+### Effective Values
+
+There are instances where a service will allocate, generate, or calculate a
+value if the client chooses not to specify one. For example: a client creates a
+virtual machine without specifying a static IP address for the virtual machine
+to be available on. Such a scenario is opting into dynamic IP address
+allocation.
+
+Some examples of these types of fields are ones that are:
+
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with an effective value ***must*** be expressed as two fields in
+the API:
+* a mutable field that can be optionally set by the user and **must not** be
+modified by the service
+* an `OUTPUT_ONLY` field that records the effective value decided on by the
+service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string effective_ip_address = 5 [
+    (google.api.field_behavior) = OUTPUT_ONLY
+  ];
+}
+```
+
+#### Naming
+
+Effective values **must** be named by prefixing `effective_` to the mutable
+field's name.
+
+### User-Specified Fields
+
+For user-specified fields, the value in response from the service **must** be
+the same as provided by the create or update request. For string fields this
+means returning the value unchanged, with two exceptions: 
+
+* When a string field has a data type annotation, a normalized string that
+  represents the given value **may** be returned.
+* When a string field has a default value indicated by the `default` keyword,
+  that default value **may** be returned when the given value is empty or
+  missing.
+
+### Normalizations
+
+A field that is normalized by the service **must** be annotated with the
+`google.api.field` extension. The allowed set of normalizations includes
+the following formats:
+
+* uuid
+* ipv4
+* ipv6
+* email
+
+Normalizations on fields **must** be described using the `google.api.field`
```

Comment:
```suggestion
Normalizations on fields **must** be described using the `google.api.field_info`
```

We had to rename it to avoid `google.protobuf.Field` collisions/confusions.